text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+
+### New Properties: `from` and `to`
+
+I'm proposing we add `from` and `to` properties on `transition` whose values are `RouteInfo` instances representing the initial and final leafmost routes for this transition. Like all RouteInfos, these are read-only and internally immutable. They are not observable, because a  `transition` instance is never changed after creation.
+
+On an initial full-page load, the `from` property will be `null`. This creates a public API for distinguishing in-app transitions from full-page reloads.
+
+### Example: testing whether route will remain active
+
+Here's an example showing how `willTransition` can figure out if the current route will remain active after the transition:
+
+```js
+willTransition(transition) {
+  if (!this.transition.to.find(route => route.name === this.routeName)) {
+    alert("Please save or cancel your changes.");
+    transition.abort();
+  }
+}
+```
+
+### Example: parent redirecting to a fallback model
+
+Here's an example of a parent route that can redirect to a fallback model, without losing its child route:
+
+```js
+this.route('person', { path: '/person/:person_id' }, function() {
+  this.route('index');
+  this.route('detail');
+});
+
+//PersonRoute
+const fallbackPersonId = 0;
+model({ personId }, transition) {
+  return this.get('store').find('person', personId).catch(err => {
+    this.replaceWith(transition.to.name, fallbackPersonId);
+  });
+}
+
+// If personId 5 is invalid, and the user visits /person/5/detail, they will get
+// redirected to /person/0/detail. And /person/5 will get redirected to /person/0.
+```
+
+
+### Actively discourage use of private API
+
+This RFC provides public API for doing the things people have become accustomed to doing via private API. To eliminate confusion over the correct way, we should hide all the private API away behind symbols, and provide deprecation warnings per our usual release policy around breaking "widely-used private APIs".
+
+Some of the private APIs we should mark and warn include:
+
+ - transition.state
+ - transition.params
+ - `lookup('router:main')` (should use `service:router` instead)
+
+
+## Dynamically-Scoped Variables
+
+The next section introduces general-purpose, dynamically-scoped variables. Why am I doing this in "The Router Service RFC"? Because dynamically-scoped RouteInfo has several nice capabilities that will be introduced in the subsequent section. And once we have one dynamically-scoped variable, we have the option of exposing that capability for other uses.
+
+### In the general case
+
+A [dynamically-scoped variable](https://en.wikipedia.org/wiki/Scope_(computer_science)#Dynamic_scoping) takes its value from its calling context, as opposed to its lexical context. Since dynamically-scoped variables are powerful to the point of potential danger, their syntax is intended to be appropriately verbose.
```

Comment:
@ef4 So if you have an index route template like

```
{{#outer-component}}
  {{#inner-component}}
    hello
  {{/inner-component}}
{{/outer-component}}
```

whose rendering "call stack" looks something like

```
index route template
outer-component's template
outer-component's block
inner-component's template
inner-component's block
```

then setting a dynamic variable in `outer-component's template` would make it available in everything below that: `outer-component's block`, `inner-component's template`, `inner-component's block`?